internal/guest/runtime/hcsv2/spec_devices.go

```diff
@@ -0,0 +1,133 @@
+// +build linux
+
+package hcsv2
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/Microsoft/hcsshim/internal/guest/storage/pci"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/opencontainers/runc/libcontainer/devices"
+	oci "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+const (
+	sysfsDevPathFormat = "/sys/dev/%s/%d:%d"
+
+	charType  = "char"
+	blockType = "block"
+
+	vpciDeviceIDTypeLegacy = "vpci"
+	vpciDeviceIDType       = "vpci-instance-id"
+)
+
+// addAssignedDevice goes through the assigned devices that have been enumerated
+// on the spec and updates the spec so that the correct device files can be mounted
+// into the resulting container by the runtime.
+func addAssignedDevice(ctx context.Context, spec *oci.Spec) error {
+	for _, d := range spec.Windows.Devices {
+		switch d.IDType {
+		case vpciDeviceIDTypeLegacy, vpciDeviceIDType:
+			// validate that the device is ready
+			fullPCIPath, err := pci.FindDeviceFullPath(ctx, d.ID)
+			if err != nil {
+				return errors.Wrapf(err, "failed to find device pci path for device %v", d)
+			}
+			// find the device node that links to the pci path we just got
+			dev, err := devicePathFromPCIPath(fullPCIPath)
+			if err != nil {
+				return errors.Wrapf(err, "failed to find dev node for device %v", d)
+			}
+			addLinuxDeviceToSpec(ctx, dev, spec, true)
+		}
```

Comment:
We don't need to handle an invalid device type?

Reply:
I suppose at this point we'd already validated it before passing it over to the guest

Reply:
Well better question, don't need to handle the gpu case here*?

Reply:
No, gpu devices are handled through a different code path, see workload_container.go. 

We could add an invalid error here but like you said, it's already validated at this point.